### ISSUE TITLE
torch.nn.CosineSimilarity() gives different results for same tensors with fp16 and fp32 datatypes

### ISSUE THREAD

**Issue by yxchng** [pytorch/pytorch#65298]
## 🐛 Bug
torch.nn.CosineSimilarity() gives different results for same tensor with fp16 and fp32 datatypes

normalizing the tensors resolves the problem but that is not the expected behavior. they should give same results regardless of normalization

## To Reproduce

```
import torch

x = torch.randn(10,2048).cuda()
y = torch.randn(10,2048).cuda()

print(torch.nn.CosineSimilarity()(x,y).mean())

x= x.half()
y =y.half()

print(torch.nn.CosineSimilarity()(x,y).mean())
print(torch.diagonal(x @ y.T).mean())

x = torch.nn.functional.normalize(x)
y = torch.nn.functional.normalize(y)

print(torch.nn.CosineSimilarity()(x,y).mean())
print(torch.diagonal(x @ y.T).mean())
```
gives
```
tensor(-0.0104, device=‘cuda:0’)
tensor(0., device=‘cuda:0’, dtype=torch.float16)
tensor(-20.9531, device=‘cuda:0’, dtype=torch.float16)
tensor(-0.0104, device=‘cuda:0’, dtype=torch.float16)
tensor(-0.0104, device=‘cuda:0’, dtype=torch.float16)
```

## Expected behavior

torch.nn.CosineSimilarity() should give same results for same tensors with fp16 and fp32 datatypes

## Environment
Collecting environment information...
PyTorch version: 1.9.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 10 (buster) (x86_64)
GCC version: (Debian 8.3.0-6) 8.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.28

Python version: 3.7.3 (default, Jan 22 2021, 20:04:44)  [GCC 8.3.0] (64-bit runtime)
Python platform: Linux-4.19.117.bsk.4-amd64-x86_64-with-debian-10.9
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB
GPU 6: Tesla V100-SXM2-32GB
GPU 7: Tesla V100-SXM2-32GB

Nvidia driver version: 450.80.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.3
[pip3] torch==1.9.0
[pip3] torchaudio==0.8.1
[pip3] torchvision==0.10.0
[conda] Could not collect

**Comment by ptrblck** [comment]
Double post from the [forum](https://discuss.pytorch.org/t/why-torch-nn-cosinesimilarity-gives-different-results-for-half-and-full-tensor/132283/2?u=ptrblck) with discussion.
